crates/app/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+mod host_component;
+pub mod locked;
+pub mod values;
+
+use ouroboros::self_referencing;
+use serde::Deserialize;
+use spin_core::{wasmtime, Engine, EngineBuilder, StoreBuilder};
+
+use host_component::DynamicHostComponents;
+use locked::{ContentPath, LockedApp, LockedComponent, LockedComponentSource, LockedTrigger};
+
+pub use async_trait::async_trait;
+pub use host_component::DynamicHostComponent;
+pub use locked::Variable;
+
+// TODO(lann): Should this migrate to spin-loader?
+#[async_trait]
+pub trait Loader {
+    async fn load_app(&self, uri: &str) -> anyhow::Result<LockedApp>;
+
+    async fn load_module(
+        &self,
+        engine: &wasmtime::Engine,
+        source: &LockedComponentSource,
+    ) -> anyhow::Result<spin_core::Module>;
+
+    async fn mount_files(
+        &self,
+        store_builder: &mut StoreBuilder,
+        component: &AppComponent,
+    ) -> anyhow::Result<()>;
+}
+
+pub struct AppLoader {
+    inner: Box<dyn Loader + Send + Sync>,
+    dynamic_host_components: DynamicHostComponents,
+}
+
+impl AppLoader {
+    pub fn new(loader: impl Loader + Send + Sync + 'static) -> Self {
+        Self {
+            inner: Box::new(loader),
+            dynamic_host_components: Default::default(),
+        }
+    }
+
+    pub fn add_dynamic_host_component<T: Send + Sync, DHC: DynamicHostComponent>(
+        &mut self,
+        engine_builder: &mut EngineBuilder<T>,
+        host_component: DHC,
+    ) -> anyhow::Result<()> {
+        self.dynamic_host_components
+            .add_dynamic_host_component(engine_builder, host_component)
+    }
+
+    pub async fn load_app(&self, uri: String) -> Result<App> {
+        let locked = self
+            .inner
+            .load_app(&uri)
+            .await
+            .map_err(Error::LoaderError)?;
+        Ok(App {
+            loader: self,
+            uri,
+            locked,
+        })
+    }
+
+    pub async fn load_owned_app(self, uri: String) -> Result<OwnedApp> {
+        OwnedApp::try_new_async(self, |loader| Box::pin(loader.load_app(uri))).await
+    }
+}
+
+impl std::fmt::Debug for AppLoader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AppLoader").finish()
+    }
+}
+
+#[self_referencing]
+#[derive(Debug)]
+pub struct OwnedApp {
+    loader: AppLoader,
+
+    #[borrows(loader)]
+    #[covariant]
+    app: App<'this>,
+}
+
+impl std::ops::Deref for OwnedApp {
+    type Target = App<'static>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+            // We know that App's lifetime param is for AppLoader, which is owned by self here.
+            std::mem::transmute::<&App, &App<'static>>(self.borrow_app())
```

Comment:
This doesn't seem sound to me.  For example:

```rust
async fn foo(loader: impl Loader + Send + Sync + 'static, uri: String) -> Result<&'static AppLoader> {
  Ok(AppLoader::new(loader).load_owned_app(uri).await?.deref().loader)
}
```
That would return a dangling reference, wouldn't it?

Reply:
That wouldn't compile would it? `.deref()` returns a `&App<'static>` which would be borrowing from the intermediate `OwnedApp` returned from `load_owned_app`.

Reply:
Right, but then I'm immediately getting the `loader` field from the `&App<'static>`, which is of type `&'static AppLoader`, i.e. does not borrow from the temporary `OwnedApp`.  So I think it would compile, but give me a minute and I'll find out for sure.

Reply:
Ohh yeah it could be used unsafely _in this crate_.

Reply:
I should probably just drop this. It's just a bit of convenience anyway.

Reply:
My suggested change would be to replace the `Deref` impl with an inherent method in `OwnedApp` with a signature like:

```rust
fn as_app<'this>(&'this self) -> &'this App<'this>
```

Reply:
Yeah the first iteration just used `AsRef` but I got annoyed typing it while knowing that it deref _should_ be possible. I agree that it isn't worth it.

Reply:
I wonder if the [recently-stabilized GAT feature](https://github.com/rust-lang/rust/pull/96709) might help here, assuming `Deref::Target` could be generic over a lifetime.  Might be worth experimenting with sometime.

Reply:
I forgot that you can't `impl AsRef` with this generic lifetime. Ended up calling the method `borrowed` since it reads a little better than `app.as_app()...`